cDSP bringup

setotau opened the pull request:
Adds support for cDSP found on sdm660 SoCs. It may be useful for offloading some computations (maybe neural networks or so)

setotau committed:
remoteproc: qcom: pas: Add support for SDM660 CDSP

Compute DSP in SDM660 is compatible with generic cdsp_resource_init
descriptor.

Signed-off-by: Nickolay Goppen <setotau@yandex.ru>

setotau committed:
arm64: dts: qcom: sdm630/660: Add CDSP-related nodes

In order to enable CDSP support for SDM660 SoC:
 * add shared memory p2p nodes for CDSP
 * add CDSP-specific smmu node
 * add CDSP periperal image loader node

Memory region for CDSP in SDM660 occupies the same spot as
TZ buffer mem defined in sdm630.dtsi (which does not have CDSP).
In sdm660.dtsi replace buffer_mem inherited from SDM630 with
cdsp_region, which is also larger in size.

SDM636 also doesn't have CDSP, so remove inherited from sdm660.dtsi
related nodes and add buffer_mem back.

Signed-off-by: Nickolay Goppen <setotau@yandex.ru>

minlexx commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -9,6 +9,143 @@
+		};
+	};
+
+	soc@0 {
```
Add `soc: soc@0 {` to `sdm630.dtsi` and use `&soc {` here. This will also require it to  be alphabetically sorted between `&mmcc` and `&tlmm` I guess

minlexx replied:
```
$ grep -rn "soc: " ./arch/arm64/boot/dts/qcom/
./arch/arm64/boot/dts/qcom/sdm670.dtsi:608:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8350.dtsi:660:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/qdu1000.dtsi:380:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sc7280.dtsi:959:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sc8180x.dtsi:775:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sdm845.dtsi:1189:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8550.dtsi:775:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/ipq9574.dtsi:231:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8250.dtsi:940:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/qcs8300.dtsi:590:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/ipq6018.dtsi:229:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sdx75.dtsi:462:      soc: soc@0 {
./arch/arm64/boot/dts/qcom/x1e80100.dtsi:773:   soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8939.dtsi:281:                                            rpmpd_opp_svs_soc: opp3 {
./arch/arm64/boot/dts/qcom/msm8939.dtsi:471:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8750.dtsi:525:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8650.dtsi:794:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sc8280xp.dtsi:786:   soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8996.dtsi:674:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8976.dtsi:466:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm6350.dtsi:780:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8917.dtsi:454:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/qcs615.dtsi:428:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sc7180.dtsi:780:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8150.dtsi:898:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sar2130p.dtsi:463:   soc: soc@0 {
./arch/arm64/boot/dts/qcom/ipq5018.dtsi:131:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/qcm2290.dtsi:433:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm6375.dtsi:804:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/ipq8074.dtsi:120:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sa8775p.dtsi:835:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8953.dtsi:439:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8998.dtsi:813:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8994.dtsi:213:                                            rpmpd_opp_svs_soc: opp3 {
./arch/arm64/boot/dts/qcom/msm8994.dtsi:346:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm4450.dtsi:378:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/msm8916.dtsi:339:                                            rpmpd_opp_svs_soc: opp3 {
./arch/arm64/boot/dts/qcom/msm8916.dtsi:438:    soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm6115.dtsi:608:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/qcs404.dtsi:307:     soc: soc@0 {
./arch/arm64/boot/dts/qcom/sm8450.dtsi:927:     soc: soc@0 {
```

minlexx commented on arch/arm64/boot/dts/qcom/sdm636.dtsi:
```diff
@@ -16,6 +16,11 @@
+/delete-node/ &cdsp_smmu;
+/ {
+	/delete-node/ smp2p-cdsp;
+};
```
I'm not sure if those really need to be deleted with `/delete-node/` or simply disabled with `status = "disabled";`

minlexx reviewed (commented):
I think removal of CDSP from 636 should come in the same commit as enabling it on 660, because it can potentially break bisecting on 636 devices if it stops at the "add cdsp to 660" commit. Therefore it will require all the DT changes here be in one commit (smp2p, smmu and pil for 660 + removal from 636); so 1st commit for remoteproc compatible, 2nd for all DT changes.

setotau commented:
Added soc label and squashed dts-related commits into one

minlexx commented:
From lavender:
```
[   30.057312] remoteproc remoteproc2: cdsp is available
[   30.063017] remoteproc remoteproc2: powering up cdsp
[   30.066151] remoteproc remoteproc2: Booting fw image cdsp.mdt, size 7068
[   30.102115] qcom_q6v5_pas 1a300000.remoteproc: segment outside memory range
[   30.105258] remoteproc remoteproc2: can't start rproc cdsp: -22
```

segment sizes:
```
xiaomi-lavender [/lib/firmware/msm-firmware-loader/mnt/modem/image]# ll cdsp*
-rwxr-xr-x 1 root root     340 Aug  9  2019 cdsp.b00
-rwxr-xr-x 1 root root    6728 Aug  9  2019 cdsp.b01
-rwxr-xr-x 1 root root    6592 Aug  9  2019 cdsp.b02
-rwxr-xr-x 1 root root  170956 Aug  9  2019 cdsp.b03
-rwxr-xr-x 1 root root 1313340 Aug  9  2019 cdsp.b04
-rwxr-xr-x 1 root root  673852 Aug  9  2019 cdsp.b05
-rwxr-xr-x 1 root root    2064 Aug  9  2019 cdsp.b06
-rwxr-xr-x 1 root root   54611 Aug  9  2019 cdsp.b07
-rwxr-xr-x 1 root root    7068 Aug  9  2019 cdsp.mdt
-rwxr-xr-x 1 root root     403 Aug  9  2019 cdspr.jsn
```

setotau commented:
buffer_mem should be increased from 100000 to 600000. ideally, it's should be separate cdsp region, but I've used buffer_mem since it uses same address

minlexx commented:
Perhaps sdm660.dtsi deserves its own cdsp region? If there is something like that in downstream

Edit: there is (at least for lavender)
```
		cdsp_fw_region@94a00000 {
			compatible = "removed-dma-pool";
			no-map;
			reg = <0x0 0x94a00000 0x0 0x600000>;
		};
```

setotau commented:
Maybe just delete buffer_mem from sdm660.dtsi and specify cdsp region instead?

minlexx commented:
Yeah, something like

```
/* On SDM660 this region is used for cDSP */
/delete-node/ &buffer_mem;

...

/ {
	...
	reserved-memory {
		...
		cdsp_mem: cdsp-region {
			reg = <0x0 0x94a00000 0x0 0x600000>;
			no-map;
		};

```

minlexx commented:
Apprently clover-common dtsi already has override for `/delete-node/ &buffer_mem;`

https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/arch/arm64/boot/dts/qcom/sdm660-xiaomi-clover-common.dtsi#L116

That's why it worked on clover :pensive: 

I see it's also moved from to 0x94a00000 0x94b00000 there

xiaomi-jasmine downstream has the same "default" sdm660 one
```
		cdsp_fw_region@94a00000 {
			compatible = "removed-dma-pool";
			no-map;
			reg = <0x00 0x94a00000 0x00 0x600000>;
		};
```
same for xiaomi-jason
same for 660-blackeberries https://github.com/tim-ecoder/blackberry-android-linux-kernel/blob/e01a1e2e91b95af41d44e424f5cda5e7b0021e34/arch/arm64/boot/dts/qcom/sdm660.dtsi#L378-L382

setotau commented:
Replaced buffer_mem with cdsp_region for sdm660

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-asus-x00td.dts:
```diff
@@ -15,7 +15,7 @@
 /delete-node/ &tz_mem;
 /delete-node/ &adsp_region;
-/delete-node/ &buffer_mem;
+/delete-node/ &cdsp_region;
```
Something's wrong here, I believe asus-x00td dts should not do that. It does define it to the same region?
```
		/* cdsp_fw_region */
		buffer_mem: buffer_region@94a00000 {
			reg = <0x0 0x94a00000 0x0 0x600000>;
			no-map;
		};
```

setotau replied:
it removes this region and defines new on same address and expanded

minlexx replied:
I think, maybe we should just drop both `/delete-node/ &buffer_mem;` and the `buffer_mem: buffer_region@94a00000 { ... }` from x00td dts? If it will be inherited in this form from sdm660.dtsi

setotau committed:
fixup! arm64: dts: qcom: sdm660: Add support for Xiaomi Mi Pad 4 / PLUS

minlexx commented:
in ~50% of boots adding cbs to smmu fails with (full log - https://paste.sr.ht/~minlexx/e1b8b62b49ace92b3ca3639ac6529d5e6dba0c07) :
```
[   33.528568] remoteproc remoteproc2: remote processor cdsp is now up
[   33.536938] Bluetooth: hci0: setting up wcn399x
[   33.568300] qcom,apr 15700000.remoteproc:glink-edge.apr_audio_svc.-1.-1: Adding APR/GPR dev: aprsvc:service:4:3
[   33.573835] qcom,apr 15700000.remoteproc:glink-edge.apr_audio_svc.-1.-1: Adding APR/GPR dev: aprsvc:service:4:4
[   33.582708] qcom,apr 15700000.remoteproc:glink-edge.apr_audio_svc.-1.-1: Adding APR/GPR dev: aprsvc:service:4:7
[   33.584265] qcom,fastrpc 1a300000.remoteproc:glink-edge.fastrpcglink-apps-dsp.-1.-1: no reserved DMA memory for FASTRPC
[   33.587634] qcom,apr 15700000.remoteproc:glink-edge.apr_audio_svc.-1.-1: Adding APR/GPR dev: aprsvc:service:4:8
[   33.595854] qcom_sns_reg.qcom-sns-reg qcom_common.sns-reg.1: got request for unmapped group id=2691
[   33.596699] qcom_sns_reg.qcom-sns-reg qcom_common.sns-reg.1: got request for unmapped group id=3050
[   33.598371] platform 1a300000.remoteproc:glink-edge:fastrpc:compute-cb@5: Adding to iommu group 4
[   33.634962] qcom_sns_reg.qcom-sns-reg qcom_common.sns-reg.1: got request for unmapped group id=2990
[   33.636419] SError Interrupt on CPU0, code 0x00000000bf000000 -- SError
[   33.636436] CPU: 0 UID: 0 PID: 2066 Comm: (udev-worker) Tainted: G        W           6.15.3-sdm660 #256 PREEMPT 
[   33.636445] Tainted: [W]=WARN
[   33.636457] Hardware name: Xiaomi Redmi Note 7 (Tianma) (DT)
[   33.636467] pstate: 80000005 (Nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   33.636487] pc : arm_smmu_write_context_bank+0x2f8/0x5e0
[   33.636512] lr : arm_smmu_init_domain_context+0x500/0x768
[   33.636520] sp : ffff800086623230
[   33.636522] x29: ffff800086623230 x28: ffff12714645a700 x27: 0000000000000000
[   33.636533] x26: 0000000000000000 x25: ffff127143823480 x24: 0000000000000000
[   33.636542] x23: ffff12714645a720 x22: ffff127143823480 x21: 0000000000000001
[   33.636550] x20: 0000000000000000 x19: ffff1271418ff680 x18: 0000000000000000
[   33.636559] x17: 00000000000009bf x16: 0000000000000000 x15: 0000000040201000
[   33.636567] x14: 0000000000000000 x13: 0000000000000000 x12: 0000002400000024
[   33.636575] x11: ffff12714151d410 x10: ffffdc34bafbbf30 x9 : ffffdc34ba9cd630
[   33.636584] x8 : 0000000000002000 x7 : ffffdc34bafbe410 x6 : ffff8000812c0000
[   33.636593] x5 : ffffdc34bafbe410 x4 : 0000000000000000 x3 : 00000000f404ff44
[   33.636601] x2 : ffff8000812c0000 x1 : 0000000000000020 x0 : 000000000000000c
[   33.636611] Kernel panic - not syncing: Asynchronous SError Interrupt
[   33.636616] CPU: 0 UID: 0 PID: 2066 Comm: (udev-worker) Tainted: G        W           6.15.3-sdm660 #256 PREEMPT 
[   33.636623] Tainted: [W]=WARN
[   33.636625] Hardware name: Xiaomi Redmi Note 7 (Tianma) (DT)
[   33.636629] Call trace:
[   33.636632]  show_stack+0x34/0x98 (C)
[   33.636642]  dump_stack_lvl+0x60/0x80
[   33.636651]  dump_stack+0x18/0x24
[   33.636659]  panic+0x174/0x388
[   33.636666]  nmi_panic+0x90/0x98
[   33.636675]  arm64_serror_panic+0x6c/0x90
[   33.636683]  do_serror+0x44/0x80
[   33.636691]  el1h_64_error_handler+0x30/0x48
[   33.636700]  el1h_64_error+0x6c/0x70
[   33.636707]  arm_smmu_write_context_bank+0x2f8/0x5e0 (P)
[   33.636715]  arm_smmu_init_domain_context+0x500/0x768
[   33.636723]  arm_smmu_attach_dev+0x60/0x1d0
[   33.636731]  __iommu_attach_device+0x2c/0x118
[   33.636739]  __iommu_device_set_domain+0x78/0xe0
[   33.636745]  __iommu_group_set_domain_internal+0x68/0x148
[   33.636753]  iommu_setup_default_domain+0x2dc/0x488
[   33.636761]  __iommu_probe_device+0x344/0x4c0
[   33.636768]  iommu_probe_device+0x3c/0x88
[   33.636775]  iommu_bus_notifier+0x38/0x88
[   33.636782]  notifier_call_chain+0x80/0x148
[   33.636790]  blocking_notifier_call_chain+0x4c/0x90
[   33.636798]  bus_notify+0x44/0x68
[   33.636807]  device_add+0x584/0x7c8
[   33.636814]  of_device_add+0x5c/0x80
[   33.636824]  of_platform_device_create_pdata+0x9c/0x148
[   33.636833]  of_platform_bus_create+0x1c4/0x4d8
[   33.636842]  of_platform_populate+0x84/0x178
[   33.636851]  fastrpc_rpmsg_probe+0x230/0x388 [fastrpc]
[   33.636871]  rpmsg_dev_probe+0xd0/0x200
[   33.636880]  really_probe+0xc8/0x3a0
[   33.636888]  __driver_probe_device+0x84/0x160
[   33.636897]  driver_probe_device+0x44/0x130
[   33.636906]  __driver_attach+0xcc/0x208
[   33.636915]  bus_for_each_dev+0x84/0x100
[   33.636923]  driver_attach+0x2c/0x40
[   33.636932]  bus_add_driver+0x118/0x250
[   33.636939]  driver_register+0x70/0x138
[   33.636950]  __register_rpmsg_driver+0x24/0x40
[   33.636957]  fastrpc_init+0x48/0xa0 [fastrpc]
[   33.636968]  do_one_initcall+0x60/0x298
[   33.636977]  do_init_module+0x5c/0x228
[   33.636986]  load_module+0x1a64/0x2038
[   33.636994]  __do_sys_init_module+0x178/0x228
[   33.637002]  __arm64_sys_init_module+0x24/0x40
[   33.637009]  invoke_syscall+0x50/0x120
[   33.637017]  el0_svc_common.constprop.0+0xc8/0xf0
[   33.637024]  do_el0_svc+0x24/0x38
[   33.637031]  el0_svc+0x30/0xd0
[   33.637038]  el0t_64_sync_handler+0x10c/0x138
[   33.637045]  el0t_64_sync+0x198/0x1a0
[   33.637059] SMP: stopping secondary CPUs
[   33.845311] Kernel Offset: 0x5c343a000000 from 0xffff800080000000
[   33.845315] PHYS_OFFSET: 0xffffed8f00000000
[   33.845318] CPU features: 0x0000,00001268,01000400,8200420b
[   33.845323] Memory Limit: none
```

quic-kdybcio committed:
dt-bindings: clock: gcc-sdm660: Add LPASS/CDSP vote clocks/GDSCs

Add defines for the missing clocks, which are required to power up the
related remote processors.

Signed-off-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>

quic-kdybcio committed:
clk: qcom: gcc-sdm660: Add missing LPASS/CDSP vote clocks

For the proper functioning of SMMUs related to the audio/compute DSPs,
it makes sense that the clocks and power domains they rely on for
communication should be online.

Add the vote clocks & GDSCs that, when enabled, ensure all such
requirements are met, through various internal mechanisms.

Signed-off-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>

minlexx commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -276,6 +307,127 @@
+
+		qcom,smem-states = <&cdsp_smp2p_out 0>;
+		qcom,smem-state-names = "stop";
+		glink-edge {
```
newline before subnodes

setotau replied:
fixed

minlexx commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -276,6 +307,127 @@
+		clock-names = "hlos1_turing";
+
+		power-domains = <&gcc HLOS1_VOTE_TURING_ADSP_GDSC>;
+		power-domain-names = "gdsc";
```
This is perhaps unneeded?

setotau replied:
Konrad said that power-domain-names is unneccessary, but power-domains is required

minlexx replied:
Yes, I', talking about `power-domain-names` specifically

minlexx commented on arch/arm64/boot/dts/qcom/sdm636.dtsi:
```diff
@@ -16,6 +16,18 @@
+/delete-node/ &cdsp_pil;
+/delete-node/ &cdsp_smmu;
+/delete-node/ &cdsp_region;
+/ {
```
newline before /

setotau replied:
fixed

minlexx reviewed (changes requested):
Attach dmesg of successful boot: https://paste.sr.ht/~setotau/c065745ed6c3c1b0190f1994213a240496bbb249

Maybe put Konrad's commits first, then yours.. dt-binding commits usually go first in series, and also I feel like clocks additions may be a separate patchset to clk, and rest 3 go to linux-arm-msm and remoteproc :thinking:

setotau committed:
dt-bindings: remoteproc: qcom: adsp: Add SDM660 CDSP compatible

Add compatible for the compute DSP remoteproc found in SDM660.

Signed-off-by: Nickolay Goppen <setotau@yandex.ru>

minlexx commented on arch/arm64/boot/dts/qcom/sdm636.dtsi:
```diff
@@ -16,6 +16,19 @@
+
+/ {
+	/delete-node/ smp2p-cdsp;
+	reserved-memory {
```
maybe a blank line before subnodes?

setotau replied:
fixed

minlexx reviewed (approved):
Seems to boot fine after 5 reboots.

minlexx commented:
Actually there are multiple dtbs-check warnings coming from new additions to `sdm660.dtsi`:
```
sdm660-xiaomi-clover-plus.dtb: iommu@5180000 (qcom,sdm630-smmu-v2): 'anyOf' conditional failed, one must be fixed:
['hlos1_turing'] is too short
'bus' was expected
'iface' was expected
'iface-mm' was expected
[[40, 132]] is too short
```

The problem is clock-names in cdsp-smmu node saying `clock-names = "hlos1_turing";` while dt-bindings in https://elixir.bootlin.com/linux/v6.16/source/Documentation/devicetree/bindings/iommu/arm,smmu.yaml#L295 inside `anyOf:` block not describing such case.

Can we name it just "bus"? To fit in the frist descirbed case.

Or, add a new `- properties:` block describing new use case and new clock name.

Upd: [msm8998](https://elixir.bootlin.com/linux/v6.16/source/arch/arm64/boot/dts/qcom/msm8998.dtsi#L1622) calls is "bus", I'll change to "bus" and proceed

minlexx commented:
Thank you, @setotau  @quic-kdybcio ! :love_you_gesture:
